clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop3/GravitinoVirtualFileSystem.java

```diff
@@ -0,0 +1,418 @@
+    return filesetCache;
+  }
+
+  private void initializeCache(int maxCapacity, long expireAfterAccess) {
+    // Since Caffeine does not ensure that removalListener will be involved after expiration
+    // We use a scheduler with one thread to clean up expired clients.
+    this.scheduler = new ScheduledThreadPoolExecutor(1, newDaemonThreadFactory());
+
+    this.filesetCache =
+        Caffeine.newBuilder()
+            .maximumSize(maxCapacity)
+            .expireAfterAccess(expireAfterAccess, TimeUnit.MILLISECONDS)
+            .scheduler(Scheduler.forScheduledExecutorService(scheduler))
+            .removalListener(
+                (key, value, cause) -> {
+                  try {
+                    Pair<Fileset, FileSystem> pair = ((Pair<Fileset, FileSystem>) value);
+                    if (pair != null && pair.getRight() != null) pair.getRight().close();
+                  } catch (IOException e) {
+                    Logger.error("Cannot close the file system for fileset: {}", key, e);
+                  }
+                })
+            .build();
+  }
+
+  private ThreadFactory newDaemonThreadFactory() {
+    return new ThreadFactoryBuilder()
+        .setDaemon(true)
+        .setNameFormat("gvfs-cache-cleaner" + "-%d")
+        .build();
+  }
+
+  private String concatVirtualPrefix(NameIdentifier identifier) {
+    return GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX
+        + identifier.namespace().level(1)
+        + "/"
+        + identifier.namespace().level(2)
+        + "/"
+        + identifier.name();
+  }
+
+  private Path getActualPathByIdentifier(
+      NameIdentifier identifier, Pair<Fileset, FileSystem> filesetPair, Path path) {
+    String virtualPath = path.toString();
+    if (!virtualPath.startsWith(GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX)) {
+      throw new InvalidPathException(
+          String.format(
+              "Path %s doesn't start with the scheme \"%s\".",
+              virtualPath, GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX));
+    }
+    try {
+      if (checkMountsSingleFile(filesetPair)) {
+        String virtualPrefix = concatVirtualPrefix(identifier);
+        Preconditions.checkArgument(
+            virtualPath.equals(virtualPrefix),
+            "Path: %s should be same with the virtual prefix: %s, because the fileset only mounts a single file.",
+            virtualPath,
+            virtualPrefix);
+
+        return new Path(filesetPair.getLeft().storageLocation());
+      } else {
+        return new Path(
+            virtualPath.replaceFirst(
+                concatVirtualPrefix(identifier),
+                new Path(filesetPair.getLeft().storageLocation()).toString()));
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(
+          String.format("Cannot resolve path: %s to actual storage path, exception:", path), e);
+    }
+  }
+
+  @SuppressWarnings("deprecation")
```

Comment:
Why do we need this annotation?

Reply:
`FileSystem.isFile(new Path(""))` is deprecated in hadoop 3. If we use `FileSystem.getFileStatus(new Path("")).isFile`, we should firstly handle the `FileNotFoundException` when get file status. This logic already implements in the old method.

Reply:
We can handle this case in our code.

Reply:
Done